lib/saluki-components/src/transforms/dogstatsd_prefix_filter/mod.rs

```diff
@@ -0,0 +1,349 @@
+        static OUTPUTS: &[OutputDefinition] = &[OutputDefinition::default_output(DataType::Metric)];
+        OUTPUTS
+    }
+
+    async fn build(&self, _: ComponentContext) -> Result<Box<dyn Transform + Send>, GenericError> {
+        Ok(Box::new(DogstatsDPrefixFilter {
+            metric_prefix: self.metric_prefix.clone(),
+            metric_prefix_blacklist: self.metric_prefix_blacklist.clone(),
+            blocklist: Blocklist::new(&self.metric_blocklist, self.metric_blocklist_match_prefix),
+        }))
+    }
+}
+
+impl MemoryBounds for DogstatsDPrefixFilterConfiguration {
+    fn specify_bounds(&self, builder: &mut MemoryBoundsBuilder) {
+        // Capture the size of the heap allocation when the component is built.
+        builder.minimum().with_single_value::<DogstatsDPrefixFilter>();
+    }
+}
+
+#[derive(Debug, Default)]
+struct Blocklist {
+    data: Vec<String>,
+    match_prefix: bool,
+}
+
+impl Blocklist {
+    fn new(data: &[String], match_prefix: bool) -> Self {
+        let mut data = data.to_owned();
+        data.sort();
+
+        // Only keep values with unique prefixes.
+        if match_prefix && !data.is_empty() {
+            let mut i = 0;
+            for j in 1..data.len() {
+                if !data[j].starts_with(&data[i]) {
+                    i += 1;
+                    data[i] = data[j].clone(); // Move item to next position
+                }
+            }
+            data.truncate(i + 1);
+        }
+
+        Blocklist { data, match_prefix }
+    }
+
+    fn contains(&self, name: &str) -> bool {
+        if self.data.is_empty() {
+            return false;
+        }
+
+        let i = self.data.binary_search(&name.to_string());
+
+        // Check for prefix match when match_prefix is true
+        if self.match_prefix {
+            let index = i.unwrap_or_else(|idx| idx);
+            if index > 0 && name.starts_with(&self.data[index - 1]) {
+                return true;
+            }
+        }
+
+        if let Ok(index) = i {
+            return name == self.data[index];
+        }
+
+        false
+    }
+}
+
+#[derive(Debug)]
+pub struct DogstatsDPrefixFilter {
+    metric_prefix: String,
+
+    metric_prefix_blacklist: Vec<String>,
+
+    blocklist: Blocklist,
+}
+
+impl DogstatsDPrefixFilter {
+    fn enrich_metric(&self, mut metric: Metric) -> Option<Metric> {
+        let metric_name = metric.context().name().deref();
+
+        if self.metric_prefix.is_empty() {
+            for s in &self.blocklist.data {
+                if s == metric_name || self.blocklist.match_prefix && metric_name.starts_with(s) {
+                    return None;
+                }
+            }
+        } else {
+            // Enrich metric with prefix if prefix is allowed.
+            let new_metric_name = if self.is_excluded(metric_name) {
+                metric.context().name().clone()
+            } else {
+                MetaString::from(format!("{}{}", self.metric_prefix, metric_name))
```

Comment:
Do we need to add a separator (like a period) between the prefix and metric name? Or does the Agent expect the user to add the trailing period to the prefix directly in their configuration?

Reply:
Oops good catch. The agent adds the trailling `.` if the user doesn't already have it as that as the suffix in the prefix. Updated the code to mimic this behavior